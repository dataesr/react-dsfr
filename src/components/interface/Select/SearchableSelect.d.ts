import * as React from 'react';

export type SearchableSelectHint = string | Object | any[];

export type SearchableSelectMessageType = 'error' | 'valid';

export interface SearchableSelectOptions {
  value: string;
  label: string;
  disabled?: boolean;
  hidden?: boolean;
}

export interface SearchableSelectProps {
  className?: string;
  disabled?: boolean;
  hint?: SearchableSelectHint;
  id?: string;
  label?: string;
  message?: string;
  messageType?: SearchableSelectMessageType;
  name?: string;
  placeholder?: string;
  onChange?: (...args: any[])=>any;
  onTextChange?: (...args: any[])=>any;
  onBlur?: (...args: any[])=>any;
  onFocus?: (...args: any[])=>any;
  onKeyDown?: (...args: any[])=>any;
  options: SearchableSelectOptions[];
  selected?: string;
  required?: boolean;
  filter?: (...args: any[])=>any;
}

declare const SearchableSelect: React.FC<SearchableSelectProps>;

export default SearchableSelect;
