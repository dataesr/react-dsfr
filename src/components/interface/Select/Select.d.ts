import * as React from 'react';
import { SelectWrapperProps } from './SelectWrapper';

export type SelectMessageType = "error" | "valid";

export interface SelectOptions {
    value: string;
    label: string;
    disabled?: boolean;
    hidden?: boolean;
}

export type SelectSelected = React.ReactElement<any> | string;

export interface SelectProps{
    id?: string;
    message?: string;
    messageType?: SelectMessageType;
    options: SelectOptions[];
    selected?: SelectSelected;
    className?: string | Object | any[];
    hint?: string | Object | any[];
    label?: string;
    required?: boolean;
    onChange?: (...args: any[])=>any;
}

declare const Select: React.FC<SelectProps>;

export default Select;

