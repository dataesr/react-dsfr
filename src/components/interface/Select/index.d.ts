import SearchableSelect from "./SearchableSelect"
import Select from "./Select"
import SelectWrapper from "./SelectWrapper"

export { SearchableSelect, Select, SelectWrapper }
