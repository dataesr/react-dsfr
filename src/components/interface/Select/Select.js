import React from 'react';
import PropTypes from 'prop-types';
import { v4 as uuidv4 } from 'uuid';
import classNames from 'classnames';

import '@gouvfr/dsfr/dist/css/selects.min.css';

/**
 *
 * @visibleName Select
 */
const Select = ({
  className,
  disabled,
  hint,
  id,
  label,
  message,
  messageType,
  name,
  onChange,
  options,
  selected,
}) => {
  const _className = classNames('rf-select', {
    [`rf-select--${messageType}`]: messageType,
  });
  const selectId = id || uuidv4();
  const messageId = uuidv4();
  const _classNameWrapper = classNames('rf-select-group', {
    [`rf-select-group--${messageType}`]: messageType,
  }, className);

  return (
    <div className={_classNameWrapper}>
      {
      label && (
        <label className="rf-label" htmlFor={selectId} aria-describedby={messageId}>
          {label}
          {hint && <span className="rf-hint-text" id={`${selectId}-desc-hint`}>{hint}</span>}
        </label>
      )
      }
      <select
        className={_className}
        data-testid="select"
        disabled={disabled}
        id={selectId}
        name={name}
        onChange={onChange}
        value={selected}
      >
        {
          options.map((opt) => (
            <option
              disabled={opt.disabled || false}
              hidden={opt.hidden || false}
              key={`${selectId}-${uuidv4()}`}
              value={opt.value}
            >
              {opt.label}
            </option>
          ))
        }
      </select>
      {(message && messageType) && <p id={messageId} className={`rf-${messageType}-text`}>{message}</p>}
    </div>
  );
};

Select.defaultProps = {
  className: '',
  disabled: false,
  hint: '',
  id: null,
  label: '',
  message: '',
  messageType: '',
  name: null,
  onChange: () => {},
  selected: '',
};

Select.propTypes = {
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array,
  ]),
  disabled: PropTypes.bool,
  hint: PropTypes.string,
  id: PropTypes.string,
  label: PropTypes.string,
  message: PropTypes.string,
  messageType: PropTypes.oneOf(['error', 'valid']),
  name: PropTypes.string,
  onChange: PropTypes.func,
  options: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    disabled: PropTypes.bool,
    hidden: PropTypes.bool,
  })).isRequired,
  selected: PropTypes.string,
};

export default Select;
