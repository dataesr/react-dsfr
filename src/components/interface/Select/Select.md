### Liste dÃ©roulante
La liste dÃ©roule fournit une liste dâ€™option parmi laquelle lâ€™utilisateur peut choisir. Seule la partie visible du composant est stylisÃ© : la liste dâ€™option dÃ©roulÃ©e conserve le style du navigateur.

```jsx padded
  const options = [
    {value: "", label: "- Select -", disabled: true, hidden: true}, 
    {value: "1", label: "Burgers &amp; sandwiches"}, {value: "2", label: "Build your own"}, 
    {value: "3", label: "Friendly meals"}, 
    {value: "4", label: "Save menu &amp; more"}
  ];
  <Select
    label="Label pour liste dÃ©roulante"
    options={options}
  />
```

### Ã‰tat d'erreur
L'Ã©tat dâ€™erreur est signalÃ© par un changement de couleur et par lâ€™apparition d'un message dâ€™erreur en-dessous du composant (cf. couleurs fonctionnelles ðŸ”— : le rouge est la couleur de lâ€™Ã©tat erreur).

```jsx padded
  const options = [
    {value: "", label: "- Select -", disabled: true, hidden: true}, 
    {value: "1", label: "Burgers &amp; sandwiches"}, {value: "2", label: "Build your own"}, 
    {value: "3", label: "Friendly meals"}, 
    {value: "4", label: "Save menu &amp; more"}
  ];
  <Select
    label="Label pour liste dÃ©roulante"
    options={options}
    message="Texte d'erreur obligatoire"
    messageType="error"
  />
```
### Ã‰tat de succÃ¨s

L'Ã©tat succÃ¨s est signalÃ© par un changement de couleur et par lâ€™apparition d'un message de succÃ¨s en-dessous du composant (cf. couleurs fonctionnelles ðŸ”— : le vert est la couleur de lâ€™Ã©tat erreur).
Le message et le label sont liÃ©s par leurs attribut id et aria-describedby .
```jsx padded
  const options = [
    {value: "", label: "- Select -", disabled: true, hidden: true}, 
    {value: "1", label: "Burgers &amp; sandwiches"}, {value: "2", label: "Build your own"}, 
    {value: "3", label: "Friendly meals"}, 
    {value: "4", label: "Save menu &amp; more"}
  ];
  <Select
    label="Label pour liste dÃ©roulante"
    options={options}
    message="Texte de validation"
    messageType="valid"
  />
```
### Ã‰tat inactif
L'Ã©tat inactif indique que utilisateur ne peut pas interagir avec l'Ã©lÃ©ment.
```jsx padded
  const options = [
    {value: "", label: "- Select -", disabled: true, hidden: true}, 
    {value: "1", label: "Burgers &amp; sandwiches"}, {value: "2", label: "Build your own"}, 
    {value: "3", label: "Friendly meals"}, 
    {value: "4", label: "Save menu &amp; more"}
  ];
  <Select
    label="Label pour liste dÃ©roulante"
    options={options}
    disabled={true}
  />
```
### Liste dÃ©roulante avec texte dâ€™aide
Il est recommandÃ© dâ€™ajouter un texte dâ€™aide sous le libellÃ© afin de faciliter le choix de lâ€™utilisateur. Il sâ€™affiche sous le libellÃ© du champ.
```jsx padded
  const options = [
    {value: "", label: "- Select -", disabled: true, hidden: true}, 
    {value: "1", label: "Burgers &amp; sandwiches"}, {value: "2", label: "Build your own"}, 
    {value: "3", label: "Friendly meals"}, 
    {value: "4", label: "Save menu &amp; more"}
  ];
  <Select
    label="Label pour liste dÃ©roulante"
    hint="Texte de description additionnel"
    options={options}
  />
```