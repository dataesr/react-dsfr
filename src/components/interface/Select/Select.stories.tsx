import { ComponentStory } from '@storybook/react';

import Select from './Select';

// https://www.systeme-de-design.gouv.fr/elements-d-interface/composants/liste-deroulante

export default {
  title: 'Composants/Liste déroulante - Select',
  component: Select,
  parameters: {
    docs: {
      description: {
        component: 'La liste déroulante permet à un utilisateur de choisir un élément dans une liste donnée.',
      },
    },
  },
};

const Template: ComponentStory<typeof Select> = (args) => (<Select {...args} />);

export const Simple = Template.bind({});
Simple.args = {
  label: 'Label pour liste déroulante',
  options: [
    { label: 'Option 1', value: '1' },
    { label: 'Option 2', value: '2' },
    { label: 'Option 3', value: '3' },
    { label: 'Option 4', value: '4' },
  ],
};
Simple.storyName = 'Liste déroulante simple';

export const ErrorState = Template.bind({});
ErrorState.args = {
  label: 'Label pour liste déroulante',
  message: 'Texte d\'erreur obligatoire',
  messageType: 'error',
  options: [
    { label: 'Option 1', value: '1' },
    { label: 'Option 2', value: '2' },
    { label: 'Option 3', value: '3' },
    { label: 'Option 4', value: '4' },
  ],
};
ErrorState.storyName = 'État d\'erreur';

export const SuccessState = Template.bind({});
SuccessState.args = {
  label: 'Label pour liste déroulante',
  message: 'Texte de validation',
  messageType: 'valid',
  options: [
    { label: 'Option 1', value: '1' },
    { label: 'Option 2', value: '2' },
    { label: 'Option 3', value: '3' },
    { label: 'Option 4', value: '4' },
  ],
};
SuccessState.storyName = 'État de succès';

export const DeactivatedState = Template.bind({});
DeactivatedState.args = {
  disabled: true,
  label: 'Label pour liste déroulante',
  options: [
    { label: 'Option 1', value: '1' },
    { label: 'Option 2', value: '2' },
    { label: 'Option 3', value: '3' },
    { label: 'Option 4', value: '4' },
  ],
};
DeactivatedState.storyName = 'État désactivé';

export const HelpText = Template.bind({});
HelpText.args = {
  hint: 'Texte de description additionnel',
  label: 'Label pour liste déroulante',
  options: [
    { label: 'Option 1', value: '1' },
    { label: 'Option 2', value: '2' },
    { label: 'Option 3', value: '3' },
    { label: 'Option 4', value: '4' },
  ],
};
HelpText.storyName = 'Liste déroulante avec texte d\'aide';
