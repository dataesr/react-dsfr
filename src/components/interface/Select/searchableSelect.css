.select-search-input {
    position: relative;
}

.select-search-options {
    background-color: var(--grey-1000-75);
    box-shadow: 0px 0px 10px var(--grey-925);
    position: absolute;
    z-index: 10;
    width: 100%;
    max-height: 32vh;
    overflow: scroll;
    display: none;
}

.select-search-options__visible {
    display: inherit;
}

.select-search-option__disabled,
.select-search-option {
    padding: 8px 16px;
}

.select-search-option {
    cursor: pointer;
}

.select-search-option__selected,
.select-search-option:hover {
    background-color: var(--background-contrast-grey);
}

.select-search-option__disabled {
    cursor: not-allowed;
    color: var(--grey-850);
}

.select-search-option__disabled:hover {
    background-color: transparent;
}
