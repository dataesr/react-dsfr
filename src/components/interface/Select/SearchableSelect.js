import React, {
  createRef, useEffect, useRef, useState,
} from 'react';
import PropTypes from 'prop-types';
import { v4 as uuidv4 } from 'uuid';
import classNames from 'classnames';

import SelectWrapper from './SelectWrapper';
import './searchableSelect.css';
import dataAttributes from '../../../utils/data-attributes';

/**
 *
 * @visibleName SearchableSelect
 */
const SearchableSelect = ({
  id,
  messageType,
  options,
  selected,
  filter,
  onChange,
  onTextChange,
  onBlur,
  onFocus,
  onKeyDown,
  ...remainingProps
}) => {
  const selectId = useRef(id || uuidv4());
  const optionsRef = useRef([]);
  const optionsContainerRef = useRef();
  const [arrowSelected, setArrowSelected] = useState();
  const [internalLabel, setInternalLabel] = useState('');
  const [showOptions, setShowOptions] = useState(false);

  if (options.length !== optionsRef.length) {
    optionsRef.current = Array(options.length)
      .fill()
      .map((option, i) => optionsRef.current[i] || createRef());
  }

  useEffect(() => {
    selectId.current = id || uuidv4();
  }, [id]);

  useEffect(() => {
    if (selected) {
      const selectedOption = options.find((option) => option.value === selected);
      setInternalLabel(selectedOption ? selectedOption.label : '');
    } else {
      setInternalLabel('');
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [selected]);

  useEffect(() => {
    if (selected) {
      const selectedOption = options.find((option) => option.value === selected);
      setInternalLabel(selectedOption ? selectedOption.label : '');
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [options]);

  useEffect(() => {
    if (arrowSelected) {
      optionsContainerRef.current.scrollTop = Math.max(
        0,
        optionsRef.current[arrowSelected].current.offsetTop - 20,
      );
    } else {
      optionsContainerRef.current.scrollTop = 0;
    }
  }, [arrowSelected]);

  const _className = classNames('fr-select', {
    [`fr-select--${messageType}`]: messageType,
  });

  const filteredOptions = options
    .filter((option, index, arr) => filter(internalLabel, option, index, arr));

  const onTextInternalChange = (value) => {
    onTextChange(value);
    setInternalLabel(value);
  };

  const onInternalChange = (newValue, newLabel) => {
    onTextInternalChange(newLabel);
    onChange(newValue);
  };

  const onInternalFocus = (e) => {
    onInternalChange('', '');
    setShowOptions(true);
    setArrowSelected(null);
    onFocus(e);
  };

  const onInternalBlur = (e) => {
    if (filteredOptions.length === 1) {
      onInternalChange(filteredOptions[0].value, filteredOptions[0].label);
    } else {
      const foundValue = options.find((option) => option.label === internalLabel);
      if (!foundValue) {
        onTextInternalChange('');
      }
    }
    setShowOptions(false);
    onBlur(e);
  };

  const onInternalKeyDown = (e) => {
    onKeyDown(e);
    switch (e.key) {
      case 'ArrowDown':
        e.preventDefault();
        setShowOptions(true);
        if (arrowSelected === null) {
          setArrowSelected(0);
        } else if (arrowSelected < filteredOptions.filter((o) => !o.disabled).length - 1) {
          setArrowSelected(arrowSelected + 1);
        }
        break;
      case 'ArrowUp':
        e.preventDefault();
        setShowOptions(true);
        if (arrowSelected && arrowSelected > 0) {
          setArrowSelected(arrowSelected - 1);
        }
        break;
      case 'Enter':
        e.preventDefault();
        if (arrowSelected !== null) {
          const option = filteredOptions.filter((o) => !o.disabled)[arrowSelected];
          onInternalChange(option.value, option.label);
          setShowOptions(false);
        }
        break;
      default:
        setArrowSelected(null);
    }
  };

  let refCount = -1;
  return (
    <SelectWrapper
      selectId={selectId.current}
      messageType={messageType}
      {...remainingProps}
    >
      <div className="select-search-input">
        <input
          {...dataAttributes.filterAll(remainingProps)}
          id={selectId.current}
          className={_className}
          autoComplete="off"
          onChange={(e) => onTextInternalChange(e.target.value)}
          onFocus={onInternalFocus}
          onBlur={onInternalBlur}
          onKeyDown={onInternalKeyDown}
          value={internalLabel}
        />
        <div
          ref={optionsContainerRef}
          className={classNames(
            'select-search-options',
            'midlength-input',
            { 'select-search-options__visible': showOptions },
          )}
        >
          {filteredOptions.length === 0 ? (
            <div className="select-search-option__disabled">
              Aucun r√©sultat
            </div>
          ) : (
            <>
              {filteredOptions.map((option) => {
                if (!option.disabled) {
                  refCount += 1;
                }
                return (
                  <div
                    role="option"
                    aria-selected={selected === option.value}
                    ref={option.disabled ? null : optionsRef.current[refCount]}
                    className={classNames(
                      'select-search-option',
                      {
                        'select-search-option__selected': !option.disabled && refCount === arrowSelected,
                        'select-search-option__disabled': option.disabled,
                      },
                    )}
                    disabled={option.disabled || false}
                    hidden={option.hidden || false}
                    key={`${selectId}-${option.value}`}
                    value={option.value}
                    onMouseDown={() => onInternalChange(option.value, option.label)}
                  >
                    {option.label}
                  </div>
                );
              })}
            </>
          )}
        </div>
      </div>
    </SelectWrapper>
  );
};

SearchableSelect.defaultProps = {
  className: '',
  disabled: false,
  hint: '',
  id: null,
  label: '',
  message: '',
  messageType: undefined,
  name: null,
  onChange: () => {},
  onTextChange: () => {},
  onBlur: () => {},
  onFocus: () => {},
  onKeyDown: () => {},
  selected: '',
  required: false,
  filter: (
    label,
    option,
  ) => option.label.toLowerCase().includes(label.toLowerCase()),
};

SearchableSelect.propTypes = {
  className: PropTypes.string,
  disabled: PropTypes.bool,
  hint: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array,
  ]),
  id: PropTypes.string,
  label: PropTypes.string,
  message: PropTypes.string,
  messageType: PropTypes.oneOf(['error', 'valid']),
  name: PropTypes.string,
  onChange: PropTypes.func,
  onTextChange: PropTypes.func,
  onBlur: PropTypes.func,
  onFocus: PropTypes.func,
  onKeyDown: PropTypes.func,
  options: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    disabled: PropTypes.bool,
    hidden: PropTypes.bool,
  })).isRequired,
  selected: PropTypes.string,
  required: PropTypes.bool,
  filter: PropTypes.func,
};

export default SearchableSelect;
