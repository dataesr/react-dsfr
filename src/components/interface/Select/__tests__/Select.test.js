import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { v4 as uuidv4 } from 'uuid';
import Select from '..';

jest.mock('uuid', () => ({
  v4: jest.fn(),
}));

describe('<Select />', () => {
  const options = [
    { value: '1', label: 'label-1' },
    { value: '2', label: 'label-2' },
  ];
  beforeEach(() => {
    uuidv4.mockImplementationOnce(() => 'xxxxxxx');
  });
  it('renders correctly', () => {
    const component = renderer
      .create(
        <Select
          label="Label pour liste déroulante"
          options={options}
          messageType="valid"
        />,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  it('in document', () => {
    render(
      <Select
        label="Label pour liste déroulante"
        options={options}
        messageType="valid"
      />,
    );
    const search = screen.getByTestId('select');
    expect(search).toBeInTheDocument();
  });

  // test de la présence de la class rf-select
  it('has rf-select class', () => {
    const { container } = render(
      <Select
        label="Label pour liste déroulante"
        options={options}
        messageType="valid"
      />,
    );
    expect(container.getElementsByClassName('rf-select').length).toBe(1);
  });

  // test de la class d'erreur
  it('has rf-select class', () => {
    const { container } = render(
      <Select
        label="Label pour liste déroulante"
        options={options}
        messageType="error"
      />,
    );
    expect(container.getElementsByClassName('rf-select--error').length).toBe(1);
  });

  // test de la class succès
  it('has rf-select class', () => {
    const { container } = render(
      <Select
        label="Label pour liste déroulante"
        options={options}
        messageType="valid"
      />,
    );
    expect(container.getElementsByClassName('rf-select--valid').length).toBe(1);
  });
});
