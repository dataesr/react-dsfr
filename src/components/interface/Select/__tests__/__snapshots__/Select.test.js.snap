// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> renders correctly 1`] = `
<div
  className="rf-select-group rf-select-group--valid"
>
  <label
    aria-describedby="label-1-xxx"
    className="rf-label"
    htmlFor="message-xxxxxxx"
  >
    Label pour liste d√©roulante
    
  </label>
  <select
    className="rf-select rf-select--valid"
    data-testid="select"
    disabled={false}
    id="message-xxxxxxx"
    name={null}
    onChange={[Function]}
    value=""
  >
    <option
      disabled={false}
      hidden={false}
      value="1"
    >
      label-1
    </option>
    <option
      disabled={false}
      hidden={false}
      value="2"
    >
      label-2
    </option>
  </select>
  
</div>
`;
